### ISSUE TITLE
Missing a Tilemap Viewer

### ISSUE THREAD

**Issue by undisbeliever** [devinacker/bsnes-plus#81]
bsnes-plus debugger does not have a tilemap viewer.

(starting a new issue instead of continuing the previous one)

**Comment by devinacker** [comment]
yep, it's probably the one last "vsnes-like" feature i've wanted to add.

I added this to the v04 milestone instead of v03 since I'm going to do a release sometime very soon, but this will probably be something I start working on ASAP afterward.

**Comment by undisbeliever** [comment]
I've got a basic tilemap viewer working.

You have to enter in the address/sizes manually. The PPUDebugger interface will need to be expanded to retrieve the tile size, tilemap size and bitdepth for each of the backgrounds.

I did try to write a generic tile drawing function that could be reused with the vram viewer but it was too slow and caused the occasional frame skip with large tilemaps. Writing a scanline based renderer specifically for the tilemap viewer ended up being 20% faster and I elliminated the frame skips.

This viewer uses QGraphicsScene to display the tilemap to the user. Hopefully OSX honours the FastTransformation setting and doesn't anti-alias the image scaling. **This needs to be tested on OSX and Windows 10**. If it works I will update the Vram Viewer to match.

https://github.com/undisbeliever/bsnes-plus/tree/update-debugger
![screenshot](https://cloud.githubusercontent.com/assets/65687/20337701/6a24f0c8-ac1e-11e6-85da-08632708054c.png)

**Comment by devinacker** [comment]
Wow, that was fast! (In fact, now I'm kind of considering letting this be in the v03 release too, just for the sake of completeness...)

I'll give it a shot on Windows 8.1 after work today, and I can take care of the updated PPU interface as well.

Thanks so much for your work adding some great PPU tools to the debugger, they're something I've wanted to be included since the beginning of the project.

**Comment by undisbeliever** [comment]
> I did try to write a generic tile drawing function that could be reused with the vram viewer but it was too slow and caused the occasional frame skip with large tilemaps. Writing a scanline based renderer specifically for the tilemap viewer ended up being 20% faster and eliminated the frame skips.

This is false.

After literally losing sleep over this, I decided to profile the code again and discovered that my laptop's cpu throttler was skewing my results. Turning off intel pstate and reprofiling has shown that the scanline based renderer is only slightly faster than the tile based one and not the 20% I initially saw.  

I blame my recent insomnia for this blunder. Personally, I would like to just merge the two commits together, edit history and just forget this brain-fart ever happened. But you've seen it and it would be un-sportman-like.

```
Test                  Scanline        Tile

2bpp 8x8 32x32        398391ns    337068ns
4bpp 8x8 32x32        428703ns    392454ns
8bpp 8x8 32x32        703470ns    678536ns
2bpp 8x8 64x32        708079ns    631600ns
4bpp 8x8 64x32        814451ns    747960ns
8bpp 8x8 64x32       1372784ns   1324243ns
2bpp 8x8 32x64        396854ns    336788ns
4bpp 8x8 32x64        431846ns    396295ns
8bpp 8x8 32x64        703330ns    683494ns
2bpp 8x8 64x64       1405890ns   1217033ns
4bpp 8x8 64x64       1594609ns   1454223ns
8bpp 8x8 64x64       2725453ns   2619291ns

2bpp 16x16 32x32     1464630ns   1221223ns
4bpp 16x16 32x32     1567567ns   1436261ns
8bpp 16x16 32x32     2589319ns   2593649ns
2bpp 16x16 64x32     2795568ns   2808170ns
4bpp 16x16 64x32     2815125ns   2859895ns
8bpp 16x16 64x32     5142182ns   5195954ns
2bpp 16x16 32x64     2705907ns   2872629ns
4bpp 16x16 32x64     3078496ns   3086459ns
8bpp 16x16 32x64     5150566ns   5159158ns
2bpp 16x16 64x64     5375883ns   5424047ns
4bpp 16x16 64x64     6122445ns   6149943ns
8bpp 16x16 64x64    10295680ns  10323835ns
```

**Comment by devinacker** [comment]
>I blame my recent insomnia for this blunder. Personally, I would like to just merge the two commits together, edit history and just forget this brain-fart ever happened.

Fine by me. It's your personal branch, after all. I don't think anyone else is going to be bothered if you decide to mess with it a bit before it goes upstream. There's always this issue thread for additional history, anyway.

**Comment by devinacker** [comment]
I did some work on expanding the PPU interface and utilizing it in the viewer. I haven't looked at the actual tile rendering code at all yet, but the handling of 16x16 tiles seems not quite right, like at the mission intro screen in Air Strike Patrol:

![image](https://cloud.githubusercontent.com/assets/4616298/20376412/73526668-ac54-11e6-9629-007ea6267b32.png)

I'm not sure if the fact that pseudo-hires is in use has anything to do with it. I don't remember any other uses of 16x16 tiles off the top of my head.

Anyway, this version of the viewer has been merged into my master branch. Thanks again for helping with this so quickly!

**Comment by Optiroc** [comment]
Wow, first of all – great work!

The rendering is unfortunately filtered on OSX:
<img width="643" alt="tilemap_8x" src="https://cloud.githubusercontent.com/assets/907446/20390284/7acebe08-acce-11e6-9289-d181743bfec9.png">

Even at 1x the image aren't aligned to the display's pixel grid (image zoomed 8x with nearest neighbour filtering for clarity):
<img width="384" alt="tilemap_1x_enlarged" src="https://cloud.githubusercontent.com/assets/907446/20390323/a15ac6ca-acce-11e6-8dba-f733a9b09375.png">

...which is probably only a problem on retina screens since I actually got the same results with the modified image scaling settings in my previous fix. I should look into a solution for perfect rendering on retina screens at some point.

**Comment by devinacker** [comment]
Does using QImage::scaled instead of QGraphicsView::scale (similar to the VRAM viewer fix) have any significant impact on performance here?

**Comment by undisbeliever** [comment]
> Does using QImage::scaled instead of QGraphicsView::scale (similar to the VRAM viewer fix) have any significant impact on performance here?

I definitely believe it will have an impact. QGraphicsView uses the GPU to scale and only preforms scaling on the viewable area. QImage::scaled will scale the entire image using the CPU, even if it is offscreen.

Secondly I was thinking of adding optional grid lines to the Tilemap Viewer and that is definitely a lot faster and easier to code in QGraphicsView then a scaled QImage.

**Comment by undisbeliever** [comment]
> I did some work on expanding the PPU interface and utilizing it in the viewer

Darn, you did it wrong. There are many games that change the tilemap settings mid-scanline (ie, Illusion of Gaia map screen) and thus your gui would not be able to show them unless we paused emulation mid scanline (which we can't do yet).

Thanks for the interface tho, I've coded the TilemapViewer and the TilemapRenderer to how I initially deigned it.
Could you please merge this in quickly before everyone else updates their forks to match yours. I apologise for my tone but I would be annoying to have a dud commit in the tree.

https://github.com/undisbeliever/bsnes-plus/tree/update-debugger
![screenshot](https://cloud.githubusercontent.com/assets/65687/20396919/8381b89c-ad33-11e6-8bdd-84de76664160.png)

---

> but the handling of 16x16 tiles seems not quite right, like at the mission intro screen in Air Strike Patrol. ... I don't remember any other uses of 16x16 tiles off the top of my head.

I wrote a test ROM that uses 16x16 tiles and found the issue, it has been fixed.

I've also added Mode0 and Mode7 support.

**Comment by devinacker** [comment]
>There are many games that change the tilemap settings mid-scanline (ie, Illusion of Gaia map screen) and thus your gui would not be able to show them unless we paused emulation mid scanline (which we can't do yet).

I'm aware of that. It's something I did spent a bit of time thinking of an ideal solution for, but the viewer in its current state was mostly intended to test the PPU interface.

>Could you please merge this in quickly before everyone else updates their forks to match yours. I apologise for my tone but I would be annoying to have a dud commit in the tree.

I'll do so when I have time later today (after reverting my own changes to the viewer because I'd rather do that than worry about rewriting history and interfering with anybody's forks).

**Comment by undisbeliever** [comment]
> I'll do so when I have time later today (after reverting my own changes to the viewer because I'd rather do that than worry about rewriting history and interfering with anybody's forks).

Look, I apologise for splitting the tree in the first place. I've been insomniac for the past few days and looking back it's causing the stress-panic irrational behaviour.

I'll accept whatever you do to my branch, just as long as it gets committed in the end.

Once the Tilemap Viewer is done, I'm going to take a break from programming and somehow unstress myself.

**Comment by devinacker** [comment]
Hey man, no worries. I probably should have just sent you a pull request in the first place instead of pushing it as-is to my master branch. Sorry if I caused you any stress by doing so.

I've been focusing on finally getting a new release out the door, especially with these exciting new features being added recently, so I may have rushed things a bit here. Nonetheless, I appreciate the effort you've put into improving the project, but be sure not to overwork yourself :)

**Comment by devinacker** [comment]
Alright, merged your better version of the viewer into master (again). I think there's some other stuff that could be added (scrolling info, viewing individual tile properties etc.) but I'll probably do it myself at some point after the next release is finished.

Thanks again for the help.
